pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ComputeAIEmbeddingsStep.java

```diff
@@ -0,0 +1,174 @@
+    if (!isSchemaCompatible(transformContext)) {
+      return;
+    }
+
+    Map<String, String> values = new HashMap<>();
+    collectFieldValuesFromKey(fields, transformContext, values);
+    collectFieldValuesFromValue(fields, transformContext, values);
+    if (values.isEmpty()) {
+      transformContext
+          .getContext()
+          .getLogger()
+          .debug("Skipping message, empty values for configured fields");
+      return;
+    }
+    final String text =
+        fields
+            .stream()
+            .map(f -> values.get(f))
+            .filter(Objects::nonNull)
+            .collect(Collectors.joining(" "));
+
+    final List<Double> embeddings = embeddingsService.computeEmbeddings(List.of(text)).get(0);
+    applyEmbeddingsToRecord(transformContext, embeddings);
+  }
+
+  private void applyEmbeddingsToRecord(TransformContext transformContext, List<Double> embeddings) {
+    GenericRecord avroRecord = (GenericRecord) transformContext.getValueObject();
+    Schema avroSchema = avroRecord.getSchema();
+    Schema modified =
+        valueSchemaCache.computeIfAbsent(
+            avroSchema,
+            schema -> {
+              AtomicBoolean fieldExists = new AtomicBoolean(false);
+              final List<Schema.Field> newFields =
+                  avroSchema
+                      .getFields()
+                      .stream()
+                      .peek(
+                          f -> {
+                            if (f.name().equals(embeddingsFieldName)) {
+                              fieldExists.set(true);
+                            }
+                          })
+                      .map(
+                          f ->
+                              new Schema.Field(
+                                  f.name(), f.schema(), f.doc(), f.defaultVal(), f.order()))
+                      .collect(Collectors.toList());
+              if (!fieldExists.get()) {
+                newFields.add(
+                    new Schema.Field(
+                        embeddingsFieldName,
+                        Schema.createArray(Schema.create(Schema.Type.DOUBLE)),
+                        "embeddings",
+                        List.of()));
+              }
+              return Schema.createRecord(
+                  avroSchema.getName(),
+                  avroSchema.getDoc(),
+                  avroSchema.getNamespace(),
+                  avroSchema.isError(),
+                  newFields);
+            });
+
+    GenericRecord newRecord = new GenericData.Record(modified);
+    for (Schema.Field field : modified.getFields()) {
+      if (field.name().equals(embeddingsFieldName)) {
+        newRecord.put(field.name(), embeddings);
+      } else {
+        newRecord.put(field.name(), avroRecord.get(field.name()));
+      }
+    }
+    if (avroRecord != newRecord) {
+      transformContext.setValueModified(true);
+    }
+    transformContext.setValueObject(newRecord);
```

Comment:
Don't you need to configure the schema ?

Reply:
It's handled in the `TransformContext#send` method. Do you think we need additional logic ?